### ISSUE TITLE
m3u8 resource not work on iphone12

### ISSUE THREAD

**Issue by suconghou** [naver/egjs-view360#355]
mp4 url is ok , but m3u8 url not work , it maybe a IOS bug , maybe we can try render video as canvas replace video element.

**Comment by WoodNeck** [comment]
Can you share the code using the video with panoviewer?
I've heard that m3u8 requires MIME type attribute so maybe that's the cause.
- https://stackoverflow.com/a/23225645

**Comment by suconghou** [comment]
https://codepen.io/suconghou/pen/abBKdEm

it is not about the MIME type , three.js I have also tried, maybe the webgl api on iphone12 cannot get data from m3u8 video element correctly.

**Comment by WoodNeck** [comment]
I think you can use [video.js](https://videojs.github.io/videojs-contrib-hls/) with our panoviewer.
Here's some demo code for it:
- https://codepen.io/woodneck/pen/dyOKGBm
- https://naver.github.io/egjs-view360/examples/panoviewer/etc/videojs.html

**Comment by suconghou** [comment]
@WoodNeck  did you try https://codepen.io/woodneck/pen/dyOKGBm on iphone12 , all the video element of course plays well ( both my example and your example ) , the main point is the rendered webgl canvas not correct

**Comment by WoodNeck** [comment]
@suconghou 
Sorry, I misunderstood as your example didn't work on my env.
I've just tested it on my iPad, and found it couldn't load video texture. (also not triggering ready event)
I agree that this is an iOS bug, and I've also found some similar issues
- https://bugs.webkit.org/show_bug.cgi?id=179417
- https://github.com/mrdoob/three.js/issues/9754

I'll work on it soon.

**Comment by suconghou** [comment]
yes , I had works on it for days, the only way I can imagine is to use a middle canvas to fix it when the resource is m3u8

**Comment by WoodNeck** [comment]
Thanks for the suggestion, I'll try that :)

**Comment by suconghou** [comment]
is there any progress ? waiting for this .

**Comment by WoodNeck** [comment]
@suconghou Sorry but it'll take a while as I'm working on the other project 😥

**Comment by suconghou** [comment]
@WoodNeck  😥   I am going to read the code and fix it by myself , any suggestions for me how can I come to the point

**Comment by suconghou** [comment]
clone & npm install ; but can not run `npm start`
error 
```
[12:00:05] admin:egjs-view360 git:(master*) $ npm start

> @egjs/view360@3.4.3-snapshot start
> webpack-dev-server --open --host 0.0.0.0

sh: webpack-dev-server: command not found
```
after `npm i webpack-dev-server `
got 
```
[12:02:43] admin:egjs-view360 git:(master*) $ npm start               

> @egjs/view360@3.4.3-snapshot start
> webpack-dev-server --open --host 0.0.0.0

The CLI moved into a separate package: webpack-cli
Please install 'webpack-cli' in addition to webpack itself to use the CLI
-> When using npm: npm i -D webpack-cli
-> When using yarn: yarn add -D webpack-cli
node:internal/modules/cjs/loader:927
  throw err;
  ^

Error: Cannot find module 'webpack-cli/bin/config-yargs'

```

what's the version of `webpack-dev-server`, maybe you should add `webpack-dev-server`  in your package.json

**Comment by WoodNeck** [comment]
@suconghou Sorry that I can't lend some time for this issue at the moment 😭 

that `start` script is actually just leftover when we're using webpack for the build script.
We don't actually have any proper automated watch & serve script for a development environment.
For now, you can try `npm run build` for every change and `npm run demo:prebuild-latest` to apply the change.
The actual demo server is can be run by calling `demo:start` script, which will require [jekyll](https://jekyllrb.com/) so please install that before.

Or you can reference `dev` script in my other repository
- https://github.com/naver/egjs-view3d/blob/main/package.json
- https://github.com/naver/egjs-view3d/blob/main/rollup.config.dev.js

PRs welcome!

**Comment by juanuys** [comment]
I've also experienced issues with iOS Safari (versions 13.7 and 14.5, using the simulator).
Here's a pen to illustrate:

https://codepen.io/juanuys/pen/ZEewPMw

There's a TODO in the Javascript which you can uncomment, which will show the video element playing the HLS video correctly, but somehow it can be processed by PanoViewer and shown in a canvas.

PS if you get a modal asking for permission I've found that I have to click Deny for it to work.

This is what I expect to see (desktop Safari):

<img width="952" alt="Screenshot 2021-06-17 at 16 14 13" src="https://user-images.githubusercontent.com/54897285/122425212-3adc9000-cf87-11eb-8345-82f3f12ddbb5.png">

However, instead I see this green screen (iOS14.5 Safari, via simulator):

<img width="808" alt="Screenshot 2021-06-17 at 16 12 46" src="https://user-images.githubusercontent.com/54897285/122425305-4cbe3300-cf87-11eb-8f40-30a67c498248.png">

**Comment by WoodNeck** [comment]
Well, sorry for the long wait but it looks like it's an issue with iOS.
I've tried the solutions mentioned on other libraries & Stackoverflow.

Tried:
- Copying the video texture to 2D canvas. Works on all other platforms but iOS.
- As it's mostly the CORS issue, like it's said in https://stackoverflow.com/questions/43287226/hls-video-streaming-on-ios-safari. I've tried serving the m3u8 file on the local(same) network but it doesn't work either.

I guess we'll just have to wait for an iOS update

**Comment by suconghou** [comment]
I temporarily solved this months ago , there is a fork  https://github.com/suconghou/egjs-view360

It render video to canvas on iphone12, a little slower , but works .

**Comment by WoodNeck** [comment]
Forgot to mention that I've tested your fork also, it doesn't work on my env (iPad 14.7.1)

**Comment by suconghou** [comment]
1. 
```javascript

new PanoViewer(this.opt.container,
    {
        video: video,
        projectionType: "equirectangular",
        useZoom: true,
    }
);
```


2.

```
video.crossOrigin = 'anonymous'
video.setAttribute('crossOrigin', 'anonymous')
video.setAttribute('crossorigin', 'anonymous')
```

3.  add Access-Control-Allow-Origin to the m3u8 url


those works on iphone12 when I did my fork

**Comment by WoodNeck** [comment]
Some of those are new, I'll try that thanks :)

**Comment by WoodNeck** [comment]
I've tried your solution, but it still doesn't work on my iPad.

1. I served m3u8 video with Access-Control-Allow-Origin tag, used `http-server`'s --cors option
![image](https://user-images.githubusercontent.com/26213435/130409508-0c2f9009-a8c8-4350-9a18-98c607a3561d.png)
> The above screenshot is taken from Windows Edge

2. Added crossorigin property to the video element
```js
video.style.display = "none";
video.crossOrigin = 'anonymous'
video.setAttribute('crossOrigin', 'anonymous')
video.setAttribute('crossorigin', 'anonymous')

var panoViewer = new eg.view360.PanoViewer(view360Container,
  {
    video,
    useZoom: true,
    isVideo: true,
    projectionType: "equirectangular"
  }
)
```

3. Tested on your fork
![image](https://user-images.githubusercontent.com/26213435/130409955-d8260878-261f-42e2-93d1-0ce63dfbd027.png)

---

Also, for further testing, I made my PanoViewer always use 2D canvas(`_initPixelSource`) and bind texture immediately to render frames as videojs already does that before.
